stream.go

```diff
@@ -161,6 +173,16 @@ func (cs *clientStream) SendMsg(m interface{}) (err error) {
 
 func (cs *clientStream) RecvMsg(m interface{}) (err error) {
 	err = recv(cs.p, cs.codec, m)
+	defer func() {
+		// err != nil indicates the termination of the stream.
+		if EnableTracing && err != nil {
+			if err != io.EOF {
+				cs.traceInfo.tr.LazyLog(&fmtStringer{"%v", []interface{}{err}}, true)
+				cs.traceInfo.tr.SetError()
+			}
+			cs.traceInfo.tr.Finish()
+		}
```

Comment:
On Thu, Jun 18, 2015 at 6:02 PM, Qi Zhao <toqizhao@gmail.com> wrote:

>
>
> On Thu, Jun 18, 2015 at 5:35 PM, David Symonds <notifications@github.com>
> wrote:
>
>> In stream.go
>> <https://github.com/grpc/grpc-go/pull/229#discussion_r32793004>:
>>
>> > @@ -161,6 +173,16 @@ func (cs *clientStream) SendMsg(m interface{}) (err error) {
>> >
>> >  func (cs *clientStream) RecvMsg(m interface{}) (err error) {
>> >  	err = recv(cs.p, cs.codec, m)
>> > +	defer func() {
>> > +		// err != nil indicates the termination of the stream.
>> > +		if EnableTracing && err != nil {
>> > +			if err != io.EOF {
>> > +				cs.traceInfo.tr.LazyLog(&fmtStringer{"%v", []interface{}{err}}, true)
>> > +				cs.traceInfo.tr.SetError()
>> > +			}
>> > +			cs.traceInfo.tr.Finish()
>> > +		}
>>
>> I wonder if some of the fields of cs should be set to nil here
>> (independent of EnableTracing) to make a clearer panic if RecvMsg is
>> called again after it returns an error. e.g. setting cs.p = nil and cs.codec
>> = nil would probably help, since an erroneous use of a trace after
>> Finish can be hard to diagnose.
>>
> I would prefer letting clientStream keep an err status. Once it becomes
> non-nil, all the follow-up recv/send operations returns with that error
> immediately. But let' do it in a separate PR.
>
hmm, but this introduces a lock to protect err access which is bad... let
me think more about this.

> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/grpc/grpc-go/pull/229/files#r32793004>.
>>
>
>